lib/Arch.cpp

```diff
@@ -0,0 +1,441 @@
+                *struct_ptr, return_register_constraints);
+            if (mapping) {
+              // There is a valid split over registers, so add the mapping
+              return *mapping;
+            } else {
+              // Struct splitting didn't work so do RVO. Assume that the pointer
+              // to the return value resides in RAX.
+              value_declaration.reg =
+                  new remill::Register("RAX", 0, 8, 0, value_declaration.type);
+            }
+            break;
+          }
+          default: {
+            LOG(ERROR) << "Encountered an unknown return type";
+            exit(1);
+          }
+        }
+
+        return_value_declarations.push_back(value_declaration);
+      }
+    }
+  }
+
+  return return_value_declarations;
+}
+
+std::vector<anvill::ParameterDecl> X86_64_SysV::BindParameters(
+    const llvm::Function &function) {
+  std::vector<anvill::ParameterDecl> parameter_declarations;
+  auto param_names = TryRecoverParamNames(function);
+  llvm::DataLayout dl(function.getParent());
+
+  // Used to keep track of which registers have been allocated
+  std::vector<bool> allocated(parameter_register_constraints.size(), false);
+
+  // Stack position of the first argument
+  unsigned int stack_offset = 16;
+
+  for (auto &argument : function.args()) {
+    anvill::ParameterDecl declaration = {};
+    declaration.type = argument.getType();
+
+    // Try to allocate from a register. If a register is not available then
+    // allocate from the stack.
+    if (remill::Register *reg = TryRegisterAllocate(
+            *argument.getType(), allocated, parameter_register_constraints)) {
+      declaration.reg = reg;
+    } else {
+      remill::Register *mem_reg =
+          new remill::Register("RSP", stack_offset, 8, 0, argument.getType());
+
+      stack_offset += dl.getTypeAllocSize(argument.getType());
+      declaration.mem_reg = mem_reg;
+    }
+
+    // Try to get a name for the IR parameter
+    // Need to add 1 because param_names uses logical numbering, but
+    // argument.getArgNo() uses index numbering
+    declaration.name = param_names[argument.getArgNo() + 1];
+
+    parameter_declarations.push_back(declaration);
+  }
+
+  return parameter_declarations;
+}
+
+remill::Register *X86_64_SysV::BindReturnStackPointer(
+    const llvm::Function &function) {
+  // For the X86_64_SysV ABI, it is always:
+  //
+  // "return_stack_pointer": {
+  //     "offset": "8",
+  //     "register": "RSP",
+  //     "type": "L"
+  // }
+
+  auto int64_ptr_ty = llvm::IntegerType::get(function.getContext(), 64);
+  return new remill::Register("RSP", 8, 8, 0, int64_ptr_ty);
+}
+
+std::vector<ParameterDecl> X86_C::BindParameters(
+    const llvm::Function &function) {
+  std::vector<anvill::ParameterDecl> parameter_declarations;
+  auto param_names = TryRecoverParamNames(function);
+  llvm::DataLayout dl(function.getParent());
+
+  // Stack position of the first argument
+  unsigned int stack_offset = 4;
+
```

Comment:
Make sure to check for explicit structure return: https://code.woboq.org/llvm/llvm/include/llvm/IR/Function.h.html#576

If it is there, then the function's return pointer value needs to be adjusted: https://gcc.godbolt.org/z/wZSHuw

Also, check for implicit struct ret, where the return value is a non-void struct, and so you need to pass in the parameter.